File: GeneratedEntityReaderClassesPython/EntityFile.py
from NoitaStream import NoitaStream
from GeneratedClasses.GeneratedClassFinder import GeneratedClassFinder
from dataclasses import dataclass, field

@dataclass
class Entity:
    Name:str = ""
    Deleted:bool = False
    Path:str = ""
    Tags:str = ""
    X:float = 0
    Y:float = 0
    ScaleX:float = 0
    ScaleY:float = 0
    Rotation:float = 0
    Components:list = field(default_factory=list)
    Children:list["Entity"] = field(default_factory=list)

    def Read(self, s:NoitaStream) -> None:
        self.Name = s.ReadBeString()
        self.Deleted = s.ReadBool()
        self.Path = s.ReadBeString()
        self.Tags = s.ReadBeString()
        self.X = s.ReadBeFloat()
        self.Y = s.ReadBeFloat()
        self.ScaleX = s.ReadBeFloat()
        self.ScaleY = s.ReadBeFloat()
        self.Rotation = s.ReadBeFloat()
        num_components = s.ReadBeUInt32()
        for i in range(0, num_components):
            componenttypename:str = s.ReadBeString()
            component = GeneratedClassFinder.GetGeneratedClass(componenttypename)
            component.ComponentName = componenttypename
            component.Read(s)
            self.Components.append(component)

        num_children = s.ReadBeUInt32()
        for i in range(0, num_children):
            child:Entity = Entity()
            child.Read(s)
            self.Children.append(child)

    def Write(self, s:NoitaStream) -> None:
        s.WriteBeString(self.Name)
        s.WriteBool(self.Deleted)
        s.WriteBeString(self.Path)
        s.WriteBeString(self.Tags)
        s.WriteBeFloat(self.X)
        s.WriteBeFloat(self.Y)
        s.WriteBeFloat(self.ScaleX)
        s.WriteBeFloat(self.ScaleY)
        s.WriteBeFloat(self.Rotation)
        s.WriteBeUInt32(len(self.Components))
        for component in self.Components:
            s.WriteBeString(component.ComponentName)
            component.Write(s)

        s.WriteBeUInt32(len(self.Children))
        for child in self.Children:
            child.Write(s);

@dataclass
class EntityFile:
    EmptyFlag:bytes = b"" # A very strange thing. In little endian, a file with 0x00020020 is empty. A file with 0x00000002 has content in it.
    # uint32 hash size here
    Hash:str = "" # This is filled with a sort of checksum. That checksum is the filename (sans .xml and path) of an xml file with the file format specification in it.
    # uint32 number of entities here
    Entities:list[Entity] = field(default_factory=list)

    def Read(self, stream:NoitaStream) -> None:
        # Read some of the stream headers
        self.EmptyFlag = stream.ReadBytes(4)
#        stream.ReadAtLeast(EmptyFlag, EmptyFlag.Length)
        self.Hash = stream.ReadBeString()
        num_entities = stream.ReadBeUInt32()
        if self.Hash != "": 
            for i in range(0, num_entities):
                entity:Entity = Entity()
                entity.Read(stream)
                self.Entities.append(entity)

    def Write(self, stream:NoitaStream) -> None:
        if len(self.Entities) == 0:
            stream.Write(bytes([ 0, 2, 0, 32, 0, 0, 0, 0 ]))
            stream.Write(b'\0' * 0x24)
            return
        else:
            stream.Write(bytes([ 0, 0, 0, 2]))

        stream.WriteBeString(self.Hash)
        stream.WriteBeUInt32(len(self.Entities))
        for entity in self.Entities:
            entity.Write(stream)

File: GeneratedEntityReaderClasses/ManualDefines/EntityFile.cs
using System;
using System.Collections.Generic;
using System.Reflection.Metadata;
using System.Reflection.PortableExecutable;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using System.Windows.Markup;

namespace GeneratedNoitaClasses
{
    public class EntityFile
    {
        public byte[] EmptyFlag { get; set; } =  new byte[4]; // A very strange thing. In little endian, a file with 0x00020020 is empty. A file with 0x00000002 has content in it.
        // uint32 hash size here
        public string Hash { get; set; } = ""; // This is filled with a sort of checksum. That checksum is the filename (sans .xml and path) of an xml file with the file format specification in it.
        // uint32 entity number here
        public List<Entity> Entities { get; set; } = new List<Entity>();

        public void Read(NoitaStream stream)
        {
            // Read some of the stream headers
            stream.ReadAtLeast(EmptyFlag, EmptyFlag.Length);
            Hash = stream.ReadBeString();
            UInt32 num_entities = stream.ReadBeUInt32();
            if(Hash != "")
            { 
                for(int i = 0; i < num_entities; i++)
                {
                    Entity entity = new Entity();
                    entity.Read(stream);
                    Entities.Add(entity);
                }
            }
        }

        public static EntityFile Read(string path)
        {
            EntityFile ef = new EntityFile();
            ef.Read(LZFile.ReadFile(path));
            return ef;
        }

        public void Write(NoitaStream stream)
        {
            if (Entities.Count == 0)
            {
                stream.Write(new byte[] { 0, 2, 0, 32, 0, 0, 0, 0 });
                stream.Write(new byte[0x24]);
                return;
            } else
            {
                stream.Write(new byte[] { 0, 0, 0, 2});
            }

            stream.WriteBeString(Hash);
            stream.WriteBeUInt32((uint)Entities.Count);
            foreach(Entity entity in Entities)
            {
                entity.Write(stream);
            }
        }
    }

    public class Entity
    {
        public string Name { get; set; } = "";
        public bool Deleted { get; set; } = false;
        public string Path { get; set; } = "";
        public string Tags { get; set; } = "";
        public float X { get; set; }
        public float Y { get; set; }
        public float ScaleX { get; set; }
        public float ScaleY { get; set; }
        public float Rotation { get; set; }
        public List<NoitaComponentBase> Components { get; set; } = [];
        public List<Entity> Children { get; set; } = [];

        public void Read(NoitaStream s)
        {
            Name = s.ReadBeString();
            Deleted = s.ReadBool();
            Path = s.ReadBeString();
            Tags = s.ReadBeString();
            X = s.ReadBeFloat();
            Y = s.ReadBeFloat();
            ScaleX = s.ReadBeFloat();
            ScaleY = s.ReadBeFloat();
            Rotation = s.ReadBeFloat();
            UInt32 num_components = s.ReadBeUInt32();
            for (int i = 0; i < num_components; i++)
            {
                string componenttypename = s.ReadBeString();
                NoitaComponentBase component = (NoitaComponentBase)GeneratedClassFinder.GetGeneratedClass(componenttypename);
                component.ComponentName = componenttypename;
                //t.make iNoitaType
                ((iNoitaType)component).Read(s);
                Components.Add(component);
            }

            UInt32 num_children = s.ReadBeUInt32();
            for (int i = 0; i < num_children; i++)
            {
                Entity child = new Entity();
                child.Read(s);
                Children.Add(child);
            }
        }

        public void Write(NoitaStream s)
        {
            s.WriteBeString(Name);
            s.WriteBool(Deleted);
            s.WriteBeString(Path);
            s.WriteBeString(Tags);
            s.WriteBeFloat(X);
            s.WriteBeFloat(Y);
            s.WriteBeFloat(ScaleX);
            s.WriteBeFloat(ScaleY);
            s.WriteBeFloat(Rotation);
            s.WriteBeUInt32((uint)Components.Count);
            foreach(NoitaComponentBase component in Components)
            {
                s.WriteBeString(component.ComponentName);
                ((iNoitaType)component).Write(s);
            }

            s.WriteBeUInt32((uint)Children.Count);
            foreach (Entity child in Children)
            {
                child.Write(s);
            }
        }

        public class ComponentContainer
        {
            public string ComponentName { get; set; } = string.Empty;
            public required iNoitaType Component { get; set; }
        }
    }
}
